packages/confluence-adapter/src/filters/deploy_space_permissions.ts

```diff
@@ -0,0 +1,295 @@
+import _ from 'lodash'
+import { logger } from '@salto-io/logging'
+import { collections, types, values } from '@salto-io/lowerdash'
+import {
+  createChangeElementResolver,
+  definitions as definitionsUtils,
+  deployment,
+  references,
+  filters,
+  fetch as fetchUtils,
+  ChangeElementResolver,
+} from '@salto-io/adapter-components'
+import { naclCase } from '@salto-io/adapter-utils'
+import {
+  PermissionObject,
+  createPermissionUniqueKey,
+  isPermissionObject,
+  restructurePermissionsAndCreateInternalIdMap,
+  transformPermissionAndUpdateIdMap,
+} from '../definitions/transformation_utils/space'
+import { ADAPTER_NAME, PERMISSION_TYPE_NAME, SPACE_TYPE_NAME } from '../constants'
+import { Options } from '../definitions/types'
+import { UserConfig } from '../config'
+
+const log = logger(module)
+const { awu } = collections.asynciterable
+
+const permissionObjectType = new ObjectType({ elemID: new ElemID(ADAPTER_NAME, PERMISSION_TYPE_NAME) })
+
+const isPermissionObjectArray = (value: unknown[]): value is PermissionObject[] => value.every(isPermissionObject)
+
+const createPermissionMap = (permissions: PermissionObject[]): Map<string, PermissionObject> => {
+  const map = new Map<string, PermissionObject>()
+  permissions.forEach(permission => {
+    map.set(createPermissionUniqueKey(permission), permission)
+  })
+  return map
+}
+
+const calculatePermissionsDiff = (
+  change: ModificationChange<InstanceElement>,
+): { permissionsToDeleted: PermissionObject[]; permissionsToAdd: PermissionObject[] } => {
+  const beforePermissions = _.get(change.data.before.value, 'permissions')
+  const afterPermissions = _.get(change.data.after.value, 'permissions')
+  if (
+    !Array.isArray(beforePermissions) ||
+    !Array.isArray(afterPermissions) ||
+    !isPermissionObjectArray(beforePermissions) ||
+    !isPermissionObjectArray(afterPermissions)
+  ) {
+    log.warn(
+      'permissions are not in expected format: %o, skipping permissions deploy filter on space %s',
+      { beforePermissions, afterPermissions },
+      getChangeData(change).elemID.getFullName(),
+    )
+    return { permissionsToDeleted: [], permissionsToAdd: [] }
+  }
+
+  const beforePermissionsMap = createPermissionMap(beforePermissions)
+  const afterPermissionsMap = createPermissionMap(afterPermissions)
+  beforePermissionsMap.forEach(permission => {
+    const key = createPermissionUniqueKey(permission)
+    if (afterPermissionsMap.has(key)) {
+      beforePermissionsMap.delete(key)
+      afterPermissionsMap.delete(key)
+    }
+  })
+  afterPermissionsMap.forEach(permission => {
+    const key = createPermissionUniqueKey(permission)
+    if (beforePermissionsMap.has(key)) {
+      beforePermissionsMap.delete(key)
+      afterPermissionsMap.delete(key)
+    }
+  })
+  return {
+    permissionsToDeleted: Array.from(beforePermissionsMap.values()),
+    permissionsToAdd: Array.from(afterPermissionsMap.values()),
+  }
+}
```

Comment:
this can work - but I think you can simplify it into something like
```suggestion
  const overlappingPermissionKeys = new Set(_.intersection(beforePermissions.map(createPermissionUniqueKey), afterPermissions.map(createPermissionUniqueKey)))
  const removeOverlaps = (permissionList: PermissionObject[]): PermissionObject[] => permissionList.filter(permission => !overlappingPermissionKeys.has(createPermissionUniqueKey(permission))
  return {
    permissionsToDeleted: removeOverlaps(beforePermissions),
    permissionsToAdd: removeOverlaps(afterPermissions),
  }
}
```

(can also take it in different directions, there's `intersectionBy` and `differenceBy`)

Reply:
I like